[MRG] New fetcher for Megatrawls Netmats

KamalakerDadi opened the pull request:
Fixes #717 

Added:
- Fetcher
- Example
- Doc
- Tests

AlexandreAbraham commented on nilearn/datasets/func.py:
```diff
@@ -1276,3 +1276,174 @@ def fetch_mixed_gambles(n_subjects=1, data_dir=None, url=None, resume=True,
+        'd300' - Group ICA brain parcellations with dimensionality = 300
+
+        If chosen to be default, network matrices data estimated from all
+        dimensionalities ['d25', 'd50', 'd100', 'd200', 'd300'] of brian
```
Should we really handle lists of options or should we let the user do the loop on its own if he wants them all? @GaelVaroquaux

KamalakerDadi replied:
I am with giving options to choose.

AlexandreAbraham replied:
Or maybe we could just always return them all.

KamalakerDadi replied:
By default, it will return all.But, I thought it would be nice to have options to choose.

AlexandreAbraham commented on nilearn/datasets/func.py:
```diff
@@ -1276,3 +1276,174 @@ def fetch_mixed_gambles(n_subjects=1, data_dir=None, url=None, resume=True,
+        Path of the data directory. Used to force data storage in a specified
+        location.
+
+    choice_dimensionality: a string or list of strings or None, Default is None
```
`n_networks` or `n_components` would be better.

KamalakerDadi replied:
Can we stick to ```n_dimensionalities``` ? since all the strings are named as 'd'. There would be 
confusion to users.

AlexandreAbraham replied:
Well, I was thinking of removing the 'd' and make it an integer.

AlexandreAbraham commented on nilearn/datasets/func.py:
```diff
@@ -1276,3 +1276,174 @@ def fetch_mixed_gambles(n_subjects=1, data_dir=None, url=None, resume=True,
+        are fetched. For example, if chosen as a string 'd25' only data
+        corresponding to d=25 is fetched.
+
+    choice_timeseries: a string or list of strings or None, Default is None
```
Here too we should find another wording.

KamalakerDadi replied:
May be this ```timeseries```. To be consistent with the names 'ts'.

AlexandreAbraham replied:
Yes but we should not let the user choose between "ts2" or "ts3". We should find better wordings.

AlexandreAbraham commented on nilearn/datasets/func.py:
```diff
@@ -1276,3 +1276,174 @@ def fetch_mixed_gambles(n_subjects=1, data_dir=None, url=None, resume=True,
+    Znet1 = network_matrices1
+    Znet2 = network_matrices2
+
+    return Znet1, Znet2
```
You do not return a Bunch here.

AlexandreAbraham commented on nilearn/datasets/func.py:
```diff
@@ -1276,3 +1276,174 @@ def fetch_mixed_gambles(n_subjects=1, data_dir=None, url=None, resume=True,
+        network_matrices2.extend(netmats2)
+
+    Znet1 = network_matrices1
+    Znet2 = network_matrices2
```
Please load the files before returning them to the user.

KamalakerDadi replied:
Can we not stick to returning path of the text files? I see confounds in some fetchers are returned the same.

AlexandreAbraham replied:
Confounds are not loaded because the masker do it automagically. It is very likely for matrices to be fed to sklearn estimators that won't load them. But we should wait for @lesteve and @GaelVaroquaux to give their mind.

KamalakerDadi replied:
I feel returning the paths can have an option to user to look if he fetched the right files. If we returned values if user does some mistake in giving input then it would be hard to identify.

lesteve commented on nilearn/datasets/func.py:
```diff
@@ -1276,3 +1276,174 @@ def fetch_mixed_gambles(n_subjects=1, data_dir=None, url=None, resume=True,
+    Register and sign the Open Access Data Use Terms at
+    ConnectomeDB: https://db.humanconnectome.org/
+    """
+    # Data is manually uploaded in NITRC
```
is -> has been or just remove this comment. I don't think it adds a lot of value.

lesteve commented on nilearn/datasets/tests/test_func.py:
```diff
@@ -411,3 +411,44 @@ def test_fetch_mixed_gambles():
+    n_output_combinations_net2 = len(net_2)
+
+    assert_equal(expected_n_combinations, n_output_combinations_net1)
+    assert_equal(expected_n_combinations, n_output_combinations_net2)
```
Can you add a rst for the description in nilearn/datasets/description and a test that the description is correctly loaded?

KamalakerDadi replied:
Fixed

bthirion commented on nilearn/datasets/func.py:
```diff
@@ -1276,3 +1276,174 @@ def fetch_mixed_gambles(n_subjects=1, data_dir=None, url=None, resume=True,
+
+    This data can be used to predict relationships between imaging data (functional
+    connectivity) and non-imaging behavioural measures such as age, sex, education, etc.
+    The network matrices are estimated between 461 functional connectivity subjects.
```
I'd rather put "The network matrices are estimated from functional connectivity datasets of 461 subjects. "

bthirion commented on nilearn/datasets/func.py:
```diff
@@ -1276,3 +1276,174 @@ def fetch_mixed_gambles(n_subjects=1, data_dir=None, url=None, resume=True,
+        'ts2' - choice denotes the timeseries signals which were extracted using
+                mutiple spatial regression.
+        'ts3' - choice denoted the timeseries signals which were extracted using
+                eigen regression.
```
Sorry, but it is not crystal-clear to me what "eigen regression" refers to. Could you possibly give a reference here ?

lesteve commented on examples/manipulating_visualizing/plot_visualize_megatrawls_netmats.py:
```diff
@@ -0,0 +1,49 @@
+For this, we need a fetcher named as :func:`nilearn.datasets.fetch_megatrawls_netmats` in
+nilearn.datasets
+
+Please see related documentation for more details.
```
Shorten the two sentences:

See :func:`nilearn.datasets.fetch_megatrawls_netmats` documentation for more details.

lesteve commented on examples/manipulating_visualizing/plot_visualize_megatrawls_netmats.py:
```diff
@@ -0,0 +1,49 @@
+    plt.figure()
+    plt.title(title)
+    plt.imshow(netmats, interpolation="nearest")
+    plt.colorbar()
```
You probably want to use a red/blue colormap like in other examples when we plot connectivity matrices.

lesteve commented on examples/manipulating_visualizing/plot_visualize_megatrawls_netmats.py:
```diff
@@ -0,0 +1,49 @@
+# Visualization
+import matplotlib.pyplot as plt
+print(" -- Showing the matrices -- ")
+list_ = {
```
This is not a list ...

KamalakerDadi replied:
I will change to ```dict_```

AlexandreAbraham replied:
It is better to have a meaningful name for viariables. `correlation_matrices` is probably better in this case.

KamalakerDadi replied:
Noted

lesteve replied:
> It is better to have a meaningful name for viariables

+100

lesteve commented on examples/manipulating_visualizing/plot_visualize_megatrawls_netmats.py:
```diff
@@ -0,0 +1,49 @@
+
+# Visualization
+import matplotlib.pyplot as plt
+print(" -- Showing the matrices -- ")
```
Maybe "Plotting correlation matrices" instead?

KamalakerDadi replied:
Thanks I will change.

lesteve commented on nilearn/datasets/func.py:
```diff
@@ -1276,3 +1276,174 @@ def fetch_mixed_gambles(n_subjects=1, data_dir=None, url=None, resume=True,
+        location.
+
+    choice_dimensionality: a string or list of strings or None, Default is None
+        Possibile options are {'d25', 'd50', 'd100', 'd200', 'd300'}
```
Typo: Possible

lesteve commented on nilearn/datasets/func.py:
```diff
@@ -1276,3 +1276,174 @@ def fetch_mixed_gambles(n_subjects=1, data_dir=None, url=None, resume=True,
+        'd200' - Group ICA brain parcellations with dimensionality = 200
+        'd300' - Group ICA brain parcellations with dimensionality = 300
+
+        If chosen to be default, network matrices data estimated from all
```
Just use "by default" rather than "if chosen to be default". This is valid for a few other places as well.

lesteve commented on nilearn/datasets/func.py:
```diff
@@ -1276,3 +1276,174 @@ def fetch_mixed_gambles(n_subjects=1, data_dir=None, url=None, resume=True,
+        If chosen a string, network matrices estimated using particular type
+        of timeseries extraction method is fetched.
+
+    resume: boolean, Default is True
```
no capital after a comma. You have done that in multiple places.

lesteve commented on nilearn/datasets/tests/test_func.py:
```diff
@@ -411,3 +411,46 @@ def test_fetch_mixed_gambles():
+    message = ("The %s you have given '%s' is invalid. ")
+    for dim_ in ['a10', 'd15', 'd30']:
+        assert_raises_regex(ValueError,
+                            (message % ('choice_dimensionality', dim_)),
```
you don't need parentheses

lesteve commented on nilearn/datasets/tests/test_func.py:
```diff
@@ -411,3 +411,46 @@ def test_fetch_mixed_gambles():
+
+    for timeserie_ in ['tt1', 'ts4', 'st2']:
+        assert_raises_regex(ValueError,
+                            (message % ('choice_timeseries', timeserie_)),
```
don't need the outer parentheses

lesteve commented on nilearn/datasets/func.py:
```diff
@@ -1276,3 +1276,174 @@ def fetch_mixed_gambles(n_subjects=1, data_dir=None, url=None, resume=True,
+    assign_names = ['dimensionalities', 'timeseries_methods']
+    standard_inputs = [dimensionalities, timeseries_methods]
+
+    for name_, input_, assign_, standard_ in zip(names, inputs, assign_names,
```
Just curious, do the `_` at the end of the variable names carry a special meaning.

KamalakerDadi replied:
Nothing special. I will change it.

lesteve replied:
As a comment, two reasons I know of for using final underscores:
* avoid shadowing a builtin like list
* scikit-learn convention for attributes that depends on data

IMHO final underscores are not that great for readability, especially when you have `some_variable` and `some_variable_` in the same function.

lesteve commented on examples/manipulating_visualizing/plot_visualize_megatrawls_netmats.py:
```diff
@@ -0,0 +1,49 @@
+import matplotlib.pyplot as plt
+print(" -- Showing the matrices -- ")
+list_ = {
+    'Full Correlation matrices of dimensionality d=100': full_correlation_matrices_d100,
```
Why the capital C in Correlation?

KamalakerDadi commented on nilearn/datasets/func.py:
```diff
@@ -1276,3 +1276,192 @@ def fetch_mixed_gambles(n_subjects=1, data_dir=None, url=None, resume=True,
+        Path of the data directory. Used to force data storage in a specified
+        location.
+
+    choice_dimensionality: a string or list of strings or None, default is None
```
ping @AlexandreAbraham @lesteve @GaelVaroquaux @bthirion 
Any suggestions for a better naming to ```choice_dimensionality``` and ```choice_timeseries``` ?

AlexandreAbraham replied:
`choice_dimensionality` -> `n_components` with value in [25, 50, 100, 200, 300] with default ?
`choice_timeseries` -> `timeseries_regression` with value in ['multiple-spatial-regression', 'eigenregression'] with default 'eigenregression'.

KamalakerDadi replied:
Ok sounds good. Then, most of the lines need to be replaced. Present code well adapted to string. I will wait for others for a final decision.

bthirion replied:
`choice_dimensionality` -> why not `dimensionality` or even `dimension` ?
`choice_timeseries` -> `regression_mode`
I would also replace 'eigenregression' with 'eigen_regression' and 'multiple-spatial-regression' with 'multiple_spatial_regression'

KamalakerDadi commented:
I have addressed your comments @bthirion @AlexandreAbraham

AlexandreAbraham commented on examples/manipulating_visualizing/plot_visualize_megatrawls_netmats.py:
```diff
@@ -0,0 +1,49 @@
+# Visualization
+print(" -- Plotting correlation matrices -- ")
+correlation_matrices = {
+    'Full correlation matrices of dimensionality d=100': full_correlation_matrices_100,
```
We use "Correlation" rather than "Full correlation".

AlexandreAbraham commented on nilearn/datasets/func.py:
```diff
@@ -1276,3 +1276,189 @@ def fetch_mixed_gambles(n_subjects=1, data_dir=None, url=None, resume=True,
+
+    for name, check_in, assign, standard in zip(error_correcting_names, user_inputs,
+                                                assign_names, standard_variables):
+        if check_in is not None:
```
None is actually not permitted according to the docs.

AlexandreAbraham commented on nilearn/datasets/func.py:
```diff
@@ -1276,3 +1276,189 @@ def fetch_mixed_gambles(n_subjects=1, data_dir=None, url=None, resume=True,
+            elif timeseries == 'eigen_regression':
+                timeseries = 'ts3'
+            each_files_netmats1 = [(os.path.join(
+                filename + '_' + str(timeseries), 'Znet1.txt'), url, opts)]
```
Why don't you give the user the choice between `correlation` and `partial_correlation`?

AlexandreAbraham commented on nilearn/datasets/func.py:
```diff
@@ -1276,3 +1276,189 @@ def fetch_mixed_gambles(n_subjects=1, data_dir=None, url=None, resume=True,
+        network_matrices1.extend(netmats1)
+        netmats2 = _fetch_files(
+            data_dir, files_netmats2[n], resume=resume, verbose=verbose)
+        network_matrices2.extend(netmats2)
```
Could you load the matrices and not return filepaths?

AlexandreAbraham commented on nilearn/datasets/func.py:
```diff
@@ -1276,3 +1276,189 @@ def fetch_mixed_gambles(n_subjects=1, data_dir=None, url=None, resume=True,
+
+    return Bunch(Fullcorrelation=network_matrices1,
+                 Partialcorrelation=network_matrices2,
+                 description=description)
```
How is the user supposed to know which file corresponds to the timeseries / dimensionality he asked for?

KamalakerDadi replied:
For dimensionality = 25 and timeseries='eigen_regression' for example:
Output filename: 3T_Q1-Q6related468_MSMsulc_d25_ts3 where ```d25``` and ```ts3``` is what you have asked for ?

I have mentioned in the Returns doc saying that ts3 is eigen_regression and ts2 is multiple_spatial_regression. To keep it with Megatrawls naming.

And also worth to mention that we  uploaded in NITRC with these file naming.

AlexandreAbraham replied:
You know how the filename is formatted but the user does not. Plus, you have to load the matrices before returning them so the user won't know what is the original filename.

bthirion commented on nilearn/datasets/func.py:
```diff
@@ -1276,3 +1276,189 @@ def fetch_mixed_gambles(n_subjects=1, data_dir=None, url=None, resume=True,
+    Proc Natl Acad Sci USA (PNAS), 106:7209-7214, 2009.
+
+    [4] S.Smith, et al. Methods for network modelling from high quality rfMRI data.
+    Meeting of the Organization for Human Brain Mapping. 2014
```
If possible, a better reference would be welcome. I believe that this what @alexsavio pointed to in #841.

lesteve replied:
This comment hasn't been tackled yet.

lesteve commented:
This needs a rebase.

KamalakerDadi commented:
@lesteve , @AlexandreAbraham I have addressed your comments. I am loading text files and returning output as arrays. I gave option to choose  between correlation and partial correlation.

When coming to tests failures, I am not sure what's happening since the files are downloading as empty files in the tmp directory. I don't know what's causing this.  Any hints would be appreciated.

Thanks

AlexandreAbraham commented:
I'll take a quick look. I think that you should focus on region extractor PR that is, I think, very close to merge (don't fix the doc in the skimage code though).

AlexandreAbraham commented on nilearn/datasets/func.py:
```diff
@@ -1275,3 +1276,213 @@ def fetch_mixed_gambles(n_subjects=1, data_dir=None, url=None, resume=True,
+    assign_names = ['dimensionalities', 'timeseries_methods']
+    standard_variables = [dimensionalities, timeseries_methods]
+
+    for name, check_in, assign, standard in izip(error_correcting_names,
```
Detail: if you start a new line, you can fit this the `for` loop on 2 lines.

AlexandreAbraham commented on nilearn/datasets/func.py:
```diff
@@ -1275,3 +1276,213 @@ def fetch_mixed_gambles(n_subjects=1, data_dir=None, url=None, resume=True,
+            raise ValueError(message % ('matrices', matrices, output_matrices))
+        output_matrices = [matrices]
+
+    n_combinations = len(dimensionalities) * len(timeseries_methods) * len(output_matrices)
```
You don't need this.

AlexandreAbraham commented on nilearn/datasets/func.py:
```diff
@@ -1275,3 +1276,213 @@ def fetch_mixed_gambles(n_subjects=1, data_dir=None, url=None, resume=True,
+    description = _get_dataset_descr(dataset_name)
+
+    network_matrices = []
+    for n in range(n_combinations):
```
`for file_netmat in files_netmats:`

AlexandreAbraham commented on nilearn/datasets/func.py:
```diff
@@ -1275,3 +1276,213 @@ def fetch_mixed_gambles(n_subjects=1, data_dir=None, url=None, resume=True,
+    network_matrices = []
+    for n in range(n_combinations):
+        netmats = _fetch_files(
+            data_dir, files_netmats[n], resume=resume, verbose=verbose)
```
`data_dir, file_netmats, resume=resume, verbose=verbose)`

AlexandreAbraham commented on nilearn/datasets/func.py:
```diff
@@ -1275,3 +1276,213 @@ def fetch_mixed_gambles(n_subjects=1, data_dir=None, url=None, resume=True,
+            elif timeseries == 'eigen_regression':
+                timeseries = 'ts3'
+            for matrices in output_matrices:
+                dim_timeseries_matrices = os.path.join(dim_timeseries + '_' + matrices)
```
There is nothing to join here.

AlexandreAbraham commented on nilearn/datasets/func.py:
```diff
@@ -1275,3 +1276,213 @@ def fetch_mixed_gambles(n_subjects=1, data_dir=None, url=None, resume=True,
+            data_dir, files_netmats[n], resume=resume, verbose=verbose)
+        arr = csv_to_array(netmats[0])
+        network_matrices.append(arr)
+
```
Several problems in your function:
* you have 2 `for` loops: one to fetch the files, one to load them. They can be merged.
* you have 3 for loops intricated for the 3 parameters. You can factorize that using itertools
* you generate a string for each item of the dataset. That mean that, if I want to have the matrices of dimensionality 20, I'll have to parse your string, which is inconvenient, you can use 3 lists instead.
* there is no reason to call `_fetch_files` several times since you only download one file

So I would write:
```python
    dataset_name = 'Megatrawls'
    data_dir = _get_dataset_dir(dataset_name, data_dir=data_dir, verbose=verbose)
    description = _get_dataset_descr(dataset_name)

    # Generate all combinations
    dims, tseries, mats = list(zip(*list(itertools.product(dimensionalities, timeseries, matrices))))
    files = []
    ts_map = dict(multiple_spatial_regression='ts2', eigen_regression='ts3')
    mat_map = dict(correlation='Znet1.txt', partial_correlation='Znet2.txt')
    for dim, tserie, mat in zip(dims, tseries, mats):
        filepath = os.path.join('3T_Q1-Q6related468_MSMsulc_d%d_%s' % (dim, ts_map[tserie]), mat_map[mat])
        files.append((filepath, url, opts))

    # Fetch all files
    files = _fetch_files(data_dir, files, resume=resume, verbose=verbose)

    # Load them
    files = [csv_to_array(f) for f in files]

    # Build the output bunch
    return Bunch(
        description=description,
        dimensionality=dims,
        timeseries=tseries,
        matrices=mats,
        data=files)
```
A good thing can also be to replace the `data` element by `correlation` and `partial_correlation` or both when pertinent.

AlexandreAbraham commented on nilearn/datasets/tests/test_func.py:
```diff
@@ -362,3 +363,51 @@ def test_fetch_mixed_gambles():
+
+@with_setup(tst.setup_tmpdata, tst.teardown_tmpdata)
+def test_fetch_megatrawls_netmats():
+    correlations = func.fetch_megatrawls_netmats(data_dir=tst.tmpdir,
```
You can take a look at the test for Harvard Oxford fetching in test_atlas.py. Before fetching, we create a dummy XML file otherwise the function crash:
```python
    dummy = open(os.path.join(ho_dir, 'HarvardOxford-Cortical.xml'), 'w')
    dummy.write("<?xml version='1.0' encoding='us-ascii'?> "
                "<metadata>"
                "</metadata>")
    dummy.close()

    ho = atlas.fetch_atlas_harvard_oxford(target_atlas,
                                          data_dir=tst.tmpdir)
```

The problem is that your dummy file must be loadable by `csv_to_array` and empty files are not. You can, for exemple, replace the call of `csv_to_array` by `np.gen_from_txt` directly. You can also determine the filepaths that will be loaded by the function and make them non empty:

```python
    for dim in [25, 100, 200, 300]:
        dummy = open(
            os.path.join(tst.tmp_dir,
                         'Megatrawls',
                         '3T_Q1-Q6related468_MSMsulc_d%d_ts3' % dim , 'Znet2.txt'), 'w')
        dummy.write("1")
        dummy.close()

AlexandreAbraham replied:
Overriding the local version of `csv_to_array` should work too:
```python

def do_nothing(*args, **kwargs):
    return np.zeros((0,))

func.csv_to_array = do_nothing
```

lesteve commented on nilearn/datasets/tests/test_func.py:
```diff
@@ -362,3 +363,48 @@ def test_fetch_mixed_gambles():
+    for dim in [25, 100, 200, 300]:
+        files_dir = os.path.join(tst.tmpdir, 'Megatrawls', '3T_Q1-Q6related468_MSMsulc_d%d_ts3' % dim)
+        os.makedirs(files_dir)
+        dummy = open(os.path.join(files_dir, 'Znet2.txt'), 'w')
```
You can use with `open(...) as` to simplify slightly the code (you don't need the .close line)

lesteve commented on nilearn/datasets/tests/test_func.py:
```diff
@@ -362,3 +363,48 @@ def test_fetch_mixed_gambles():
+
+@with_setup(tst.setup_tmpdata, tst.teardown_tmpdata)
+def test_fetch_megatrawls_netmats():
+    for dim in [25, 100, 200, 300]:
```
Can you add a comment explaining why you need to create these mock files?

lesteve commented on nilearn/datasets/tests/test_func.py:
```diff
@@ -362,3 +363,48 @@ def test_fetch_mixed_gambles():
+                                                 timeseries=['eigen_regression'],
+                                                 matrices=['partial_correlation'])
+
+    assert_true(correlations.partial_correlation[0], np.ndarray)
```
:confused: what are you trying to test here? assert_true takes two arguments:

```
Signature: assert_true(expr, msg=None)
```

lesteve commented on nilearn/datasets/tests/test_func.py:
```diff
@@ -362,3 +363,48 @@ def test_fetch_mixed_gambles():
+
+    dimensionality = ['a10', 'd15', 'd30']
+    assert_raises_regex(ValueError,
+                        message % 'dimensionality',
```
It would be great to test that the invalid value is part of the error message. Also I would really make it obvious which one it is for people not familiar with this fetcher, i.e. call it 'invalid_input' rather than 'a10', 'multiple_sptial_regression', etc ...

lesteve commented on nilearn/datasets/tests/test_func.py:
```diff
@@ -362,3 +363,48 @@ def test_fetch_mixed_gambles():
+                        func.fetch_megatrawls_netmats,
+                        matrices=matrices)
+
+    assert_raises(TypeError, func.fetch_megatrawls_netmats,
```
Use assert_raises_regex with an error message just for explicitness

KamalakerDadi commented:
@lesteve I have addressed your comments.

KamalakerDadi commented on nilearn/datasets/tests/test_func.py:
```diff
@@ -362,3 +363,57 @@ def test_fetch_mixed_gambles():
+                                                 matrices=['partial_correlation'])
+
+    # expected number of returns sitting in ouput name correlations should be equal
+    assert_equal(len(correlations), 7)
```
Well, here I know we could test more than this. Any suggestions would be welcome.

lesteve replied:
Check each key content, e.g.:

```python
assert_equal(correlations.dimensions_partial, [25, 100, 200, 300])
assert_equal(correlations.timeseries_partial=['eigen_regression'] * 4
# etc...
```

Side-comment: correlations is not a great name, maybe just megatrawl_dataset or something similar.

lesteve commented on nilearn/datasets/func.py:
```diff
@@ -1275,3 +1277,194 @@ def fetch_mixed_gambles(n_subjects=1, data_dir=None, url=None, resume=True,
+          corresponding matrices.
+        - timeseries_partial: array of timeseries methods given to fetch its
+          corresponding matrices.
+        - Data description
```
you didn't use the name of the key here, I guess `description: ...`

lesteve commented on examples/manipulating_visualizing/plot_visualize_megatrawls_netmats.py:
```diff
@@ -0,0 +1,44 @@
+from nilearn import plotting
+
+
+def plot_mats(netmats, title):
```
That's plotting only one matrix, right? If yes, just a naming suggestion: `plot_matrix(matrix, title)`

lesteve commented on examples/manipulating_visualizing/plot_visualize_megatrawls_netmats.py:
```diff
@@ -0,0 +1,44 @@
+
+# Visualization
+print(" -- Plotting correlation matrices -- ")
+for matrices, dim, tseries in zip(
```
matrices is a plural, use `matrix` instead

lesteve commented on examples/manipulating_visualizing/plot_visualize_megatrawls_netmats.py:
```diff
@@ -0,0 +1,44 @@
+    plot_mats(matrices, title)
+
+print(" -- Plotting partial correlation matrices -- ")
+for matrices, dim, tseries in zip(
```
Idem

lesteve commented on examples/manipulating_visualizing/plot_visualize_megatrawls_netmats.py:
```diff
@@ -0,0 +1,44 @@
+print(" -- Plotting partial correlation matrices -- ")
+for matrices, dim, tseries in zip(
+        netmats.partial_correlation, netmats.dimensions_partial, netmats.timeseries_partial):
+    title = ('Partial correlation matrices of d=%d & timeseries=%s' % (dim, tseries))
```
You don't need parentheses here

lesteve commented on nilearn/datasets/func.py:
```diff
@@ -1275,3 +1277,192 @@ def fetch_mixed_gambles(n_subjects=1, data_dir=None, url=None, resume=True,
     return data
+
+
+def fetch_megatrawls_netmats(data_dir=None, dimensionality=[25, 50, 100, 200, 300],
```
You should probably use tuples for default arguments. Using mutable arguments as defaults is one of the Python gotchas. See http://docs.python-guide.org/en/latest/writing/gotchas/#mutable-default-arguments for example.

lesteve commented on nilearn/datasets/func.py:
```diff
@@ -1275,3 +1277,192 @@ def fetch_mixed_gambles(n_subjects=1, data_dir=None, url=None, resume=True,
+        of array (25, 25) and (50, 50) will be fetched.
+
+    timeseries: string or strings as list, optional
+        Valid methods in strings are ['multiple_spatial_regression', 'eigen_regression']
```
"Valid inputs" seem slightly better.

lesteve commented on nilearn/datasets/func.py:
```diff
@@ -1275,3 +1277,192 @@ def fetch_mixed_gambles(n_subjects=1, data_dir=None, url=None, resume=True,
+        If set to ['eigen_regression'], then correlation matrices estimated using
+        first principal eigen component based extraction of subject specific
+        timeseries signals will be returned.
+        For full technical details about each method. Refer to [3] [4] [5]
```
This should be a comma: "For full technical details about each method, refer to [3] [4] [5]."

lesteve commented on nilearn/datasets/func.py:
```diff
@@ -1275,3 +1277,192 @@ def fetch_mixed_gambles(n_subjects=1, data_dir=None, url=None, resume=True,
+        timeseries signals will be returned.
+        For full technical details about each method. Refer to [3] [4] [5]
+
+    matrices: string or strings as list, optional
```
I tend to use the python types, here, i.e. bool rather than boolean, str rather than string. In this case I would say:

str or list of str, optional

lesteve commented on nilearn/datasets/func.py:
```diff
@@ -1275,3 +1277,192 @@ def fetch_mixed_gambles(n_subjects=1, data_dir=None, url=None, resume=True,
+        For full technical details about each method. Refer to [3] [4] [5]
+
+    matrices: string or strings as list, optional
+        Valid output matrices in strings are ['correlation', 'partial_correlation']
```
Maybe valid "inputs" are: ...

lesteve commented on nilearn/datasets/func.py:
```diff
@@ -1275,3 +1277,192 @@ def fetch_mixed_gambles(n_subjects=1, data_dir=None, url=None, resume=True,
+        will be fetched.
+
+    resume: boolean, default is True
+        This parameter is required if a partly downloaded file is needed to be
```
"partially downloaded" seems to win the google contest against "partly downloaded"

lesteve commented on nilearn/datasets/func.py:
```diff
@@ -1275,3 +1277,192 @@ def fetch_mixed_gambles(n_subjects=1, data_dir=None, url=None, resume=True,
+          given as inputs.
+        - timeseries_correlation: array of timeseries method given as inputs.
+        - timeseries_partial: array of timeseries method given as inputs.
+        - description: Data description
```
Data -> data

lesteve commented on nilearn/datasets/func.py:
```diff
@@ -1275,3 +1277,192 @@ def fetch_mixed_gambles(n_subjects=1, data_dir=None, url=None, resume=True,
+        if not isinstance(each_input, list):
+            raise TypeError("Input given for {0} should be in list. "
+                            "You have given as single variable: {1}".format(name, each_input))
+        elif isinstance(each_input, list):
```
You can use else here. In other words there is no point using elif if your elif condition is the negation of your if condition.

lesteve commented on nilearn/datasets/func.py:
```diff
@@ -1275,3 +1277,192 @@ def fetch_mixed_gambles(n_subjects=1, data_dir=None, url=None, resume=True,
+
+    for each_input, standard, name in izip(inputs, standards, error_names):
+        if not isinstance(each_input, list):
+            raise TypeError("Input given for {0} should be in list. "
```
should be **a** list, maybe you can use `collections.Sequence` rather than list, i.e. you could argue tuple could work.

KamalakerDadi commented:
ping @lesteve @AlexandreAbraham I have addressed your comments.

KamalakerDadi commented:
@lesteve @AlexandreAbraham Any more reviews on this ?

lesteve commented on nilearn/_utils/param_validation.py:
```diff
@@ -63,3 +63,31 @@ def check_threshold(threshold, data, percentile_func, name='threshold'):
+    message = ("Invalid {0} name is given: {1}. "
+               "Please choose either of them {2}")
+
+    if isinstance(inputs, _basestring):
```
:confused: why don't you say `isinstance(inputs, collection.SequenceType)` or something like that if you want your inputs to be a list.

Also amend the error message to say "sequence" and show the type of the inputs.

lesteve commented on nilearn/_utils/param_validation.py:
```diff
@@ -63,3 +63,31 @@ def check_threshold(threshold, data, percentile_func, name='threshold'):
     return threshold
+
+
+def check_parameters_megatrawls_datasets(inputs, standards, name):
```
This doesn't seem like this is general enough to be in _utils.param_validation. Why not put it next to your fetch_megatrawls function.

KamalakerDadi replied:
You mean port this function directly into func.py or code ?

Why is it not do you think ?

lesteve replied:
Answered that IRL.

lesteve commented on nilearn/_utils/param_validation.py:
```diff
@@ -5,7 +5,7 @@
 import warnings
 
-from .compat import _basestring
+from .compat import _basestring, izip
```
You should really really really configure your editor to run pyflake8. This would have shown you that izip is unused.

lesteve commented on nilearn/datasets/tests/test_func.py:
```diff
@@ -16,6 +16,7 @@
 from . import test_utils as tst
 
 from nilearn.datasets import utils, func
+from nilearn._utils.testing import assert_raises_regex
```
This import doesn't seem to be used.

AlexandreAbraham commented on nilearn/datasets/func.py:
```diff
@@ -1279,3 +1282,189 @@ def fetch_mixed_gambles(n_subjects=1, data_dir=None, url=None, resume=True,
+    dimensions_correlation = [dims[id_c] for id_c in ids_correlation]
+    dimensions_partial = [dims[id_p] for id_p in ids_partial]
+    timeseries_correlation = [tseries[id_c] for id_c in ids_correlation]
+    timeseries_partial = [tseries[id_p] for id_p in ids_partial]
```
I find the output overly complicated. Why splitting everything between `correlation` and `partial_correlation` and not having three lists in the bunch: `dimensionality`, `timeseries` and `matrices`? I am still convinced that allowing only one value per parameter would be much simpler: this is what we already do for most of the fetchers (for harvard oxford, only one type of atlas is allowed, for aal, it's the same, only one version at a time).

AlexandreAbraham commented:
Apart from my comment about output, I'm +1.

banilo commented on examples/manipulating_visualizing/plot_visualize_megatrawls_netmats.py:
```diff
@@ -0,0 +1,51 @@
+
+
+def plot_matrix(matrix, title):
+    plt.figure()
```
``plt`` might not be defined upon function call

banilo commented on examples/manipulating_visualizing/plot_visualize_megatrawls_netmats.py:
```diff
@@ -0,0 +1,51 @@
+# and array 2 has matrix with dimensionality=100
+correlation_matrices = netmats.partial_correlation
+
+# Array of given dimensions
```
is it possible to make this comment more informative?

banilo commented on examples/manipulating_visualizing/plot_visualize_megatrawls_netmats.py:
```diff
@@ -0,0 +1,51 @@
+import matplotlib.pyplot as plt
+from nilearn import plotting
+
+for matrix, dim, tserie in zip(correlation_matrices, dimensions_partial,
```
minor comment, but ``tseries`` might be more natural

bthirion replied:
`time_series` would actually be much better

banilo commented on nilearn/datasets/description/Megatrawls.rst:
```diff
@@ -0,0 +1,56 @@
+parcellations. In total, 461 functional connectivity datasets were used.
+
+Data is available for all dimensionalities (d=25, d=50, d=100, d=200, d=300)
+of Group ICA and for all timeseries methods multiple spatial regression and
```
grammar

banilo commented on nilearn/datasets/func.py:
```diff
@@ -1279,3 +1282,189 @@ def fetch_mixed_gambles(n_subjects=1, data_dir=None, url=None, resume=True,
+        cerebellum defined by resting state functional connectivity.
+        Cerebral Cortex, 2009.
+
+    Disclaimer
```
We might want to move this section to the corresponding *.rst file

banilo commented on nilearn/datasets/tests/test_func.py:
```diff
@@ -362,3 +363,37 @@ def test_fetch_mixed_gambles():
+    # smoke test to see that files are fetched and read properly
+    # since we are loading information present in it and returning
+    # the same
+    dimensionality = [25, 100, 200, 300]
```
Aren't you missing some cases here. Dimensinoalities appear to be ``[25, 50, 100, 200, 300]``

KamalakerDadi commented:
@AlexandreAbraham Just to inform you that I addressed your comments. It would be nice to have a look whenever you have time. Thanks.
